website/blog/dec-2021/structure-chaining.md

```diff
@@ -0,0 +1,69 @@
+---
+{
+"TableOfContents": {
+"Name": "Silk.NET Structure Chaining",
+"Url": "structure-chaining.html",
+"Metadata": {
+"AuthorGitHub": "thargy",
+"DateTimeWritten": "09/12/2021 16:00",
+"PreviewImage": "images/blog/dec-2021/structure-chaining.png"
+} } }
+---
+
+# Silk.NET Structure Chaining
+
+With the release of Silk.NET 2.11 we've enhanced the usability of `Silk.NET.Vulkan` by addressing one of its pain
+points, specifically building the [Singly Linked Lists](https://en.wikipedia.org/wiki/Linked_list) (or chains) of
+structures that are frequently used to pass information to, and from, the API.
+
+Up until now, it was not unusual to have to write quite a lot of unsafe code to work with the Vulkan API. For example, to
+query device features you would need to create a chain of empty structures to be populated by the API:
+
+```csharp
+var accelerationStructureFeaturesKhr = new PhysicalDeviceAccelerationStructureFeaturesKHR
+    {SType = StructureType.PhysicalDeviceAccelerationStructureFeaturesKhr};
+var indexingFeatures = new PhysicalDeviceDescriptorIndexingFeatures
+{
+    SType = StructureType.PhysicalDeviceDescriptorIndexingFeatures,
+    PNext = &accelerationStructureFeaturesKhr
+};
+var features2 = new PhysicalDeviceFeatures2
+{
+    SType = StructureType.PhysicalDeviceFeatures2,
+    PNext = &indexingFeatures
+};
+vk.GetPhysicalDeviceFeatures2(device, &features2);
+```
+
+Not only is this quite a bit of boiler plate (and it gets worse when you actually want to pass data to the API), it is
+really easy for developers, who are not always comfortable with `unsafe` code, memory management and pointers, to
+introduce subtle bugs by passing in pointers to structures they have previously generated and are now residing on the
+.NET heap. Such structures can be moved by the Garbage Collector at the most unexpected times!
+
+Thanks to Structure Chaining, we now have 2 new ways to write the above code. The most
+useful [Managed Chaining](https://github.com/dotnet/Silk.NET/blob/main/documentation/structure-chaining/managed-chaining.md),
+makes it much easier, _and safer_, to work with Vulkan Structure Chains. For example, the above code can now be written
+as:
+
+```csharp
+using var chain = Chain.Create<PhysicalDeviceFeatures2, PhysicalDeviceDescriptorIndexingFeatures,
+    PhysicalDeviceAccelerationStructureFeaturesKHR>();
+vk.GetPhysicalDeviceFeatures2(device, features2);
```

Comment:
Where does `features2` come from?

Reply:
Oh right this is a continuation of the previous snippet. Don't mind me!

Reply:
Wait actually unresolving this as, if I understand correctly, `chain` will be empty and we're not even passing it Vulkan.

Reply:
The chain will have the three structures in it, with the `PNext` and `STypes` set correctly, but otherwise empty.  This is passed to `vk.GetPhysicalDeviceFeatures2` which populates it, at which point you can access the values in the structures using `chain.Head....` and `chain.Item1...` and `chain.Item2....`  It is the exact equivalent of the code above it.

Reply:
Ah Vulkan populates it instead of us populating it! That was what I was missing...

In any case, shouldn't it be
```suggestion
vk.GetPhysicalDeviceFeatures2(device, chain);
```
?

Reply:
Aww F*$£*$£, yeah it should say `chain`, I'll fix.

Reply:
If fixed the chain name to features2 to be consistent with above